server/src/main/kotlin/org/javacs/kt/SourcePath.kt

```diff
@@ -60,9 +63,11 @@ class SourcePath(
                 LOG.debug("Compiling {}", path?.fileName)
 
                 val (context, container) = cp.compiler.compileFile(parsed!!, allIncludingThis(), kind)
-                compiledContext = context
-                compiledContainer = container
-                compiledFile = parsed
+                parseDataWriteLock.withLock {
```

Comment:
Adding a lock at least for writing into the compiled* fields - otherwise, there are generally multiple threads writing to those, and that may cause data races, although this would be very rare.